### PR TITLE
Fixes disappearing toolbar on click in Safari.

### LINKED ISSUE #12115: Toolbars disappearing on Safari
## 📝 Provide detailed reproduction steps (if any)

1. Open the editor, and scroll up, so that the toolbar sticks on top.
2. Use any feature from the toolbar on a portion of the content (eg. insert a link, apply formatting)

### ✔️ Expected result

The feature works as expected and the toolbar stays put.

### ❌ Actual result

The toolbar disappears (probably "jumps" to the top, out of viewport).

### ❓ Possible solution

@FilipTokarski suggested this might be due to changes in the focus area.

## 📃 Other details

* Browser: Safari
* OS: MacOS
* First affected CKEditor version: 35.0.0
* Installed CKEditor plugins: …

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by mmotyczynska**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Internal: Fixes disappearing toolbar on click in Safari. Closes #12115.

---

### Additional information

Used internal commit category, since the source bug was not released yet.

### CODE REVIEW

**Comment by mmotyczynska** [comment]
@ckeditor/qa-team could you please verify if the current fix solves the issue? If possible, please check it in older/slower MacBooks as we have to find rather small value for the timeout which still works even for older machines. Currently it's set to 150ms and it works on MacBook Air (M1, 2020).

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  In Safari, go to http://localhost:8080/ckeditor5/34.2.0/examples/builds-custom/full-featured-editor.html.
2.  Focus the editor and scroll down, so the sticky toolbar appears.
3.  Click on the split button arrow, e.g. in the media embed icon.
4.  Repeat step 3.

## Current result

The toolbar disappears.

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  In Safari, go to http://localhost:8080/ckeditor5/34.2.0/examples/builds/balloon-editor.html.
2.  Select some text.
3.  Click on the split button arrow, e.g. in the media embed icon.

## Current result

The toolbar disappears.

**Comment by Reinmar** [comment]
Please check https://github.com/ckeditor/ckeditor4/blob/master/core/focusmanager.js#L64 (history of that line may give more context).

**Comment by Reinmar** [comment]
BTW, CC is missing in utils 😕 ?

**Comment by Reinmar** [comment]
> BTW, CC is missing in utils 😕 ?

Ah, right – that's due to the conditional statement. Some quick test that stubs the `env.isSafari` should be enough

**Comment by Reinmar** [comment]
I found one blocker: table dropdown doesn't manage the focus correctly. Upon opening the dropdown the focus disappears so the toolbar disappears too.

The solution here will be to do something reasonable with the focus on opening the dropdown – e.g. moving it to the first button in the grid. Right now, we don't have keyboard navigation in grids, so this change can be temporary, but we need to do something now.

Once this issue is fixed, the next priority will be to manage the focus correctly in the dropdown+splitbutton case. E.g. a second click on the arrow in the "highlight" button will indeed lose focus. That's because on the second click on the arrow we do nothing with the focus. The correct solution will be to move it correctly onto the arrow button.

**Comment by Mgsy** [comment]
Note: It seems there is a difference between clicking on an item in the toolbar by "tapping" and hard-clicking the touchpad. Also, clicking and holding the touchpad makes the toolbar disappear almost every time.

**Comment by Reinmar** [comment]
> Note: It seems there is a difference between clicking on an item in the toolbar by "tapping" and hard-clicking the touchpad. Also, clicking and holding the touchpad makes the toolbar disappear almost every time.

One thing that we need to understand first what events are fired in both cases. Apparently, Safari behaves differently in them.

**Comment by mmotyczynska** [comment]
One more issue: when you press and hold any button in toolbar for at least 200ms (or whatever we set the timeout to) then the toolbar disappears.

**Comment by mmotyczynska** [comment]
> could you please verify if the current fix solves the issue

@Mgsy did you manage to check the best value of timeout for simple cases like bold button (or other not-dropdown buttons)? We have several issues here and we have to solve them one by one, so let's start with this simple case.

**Comment by Mgsy** [comment]
> @Mgsy did you manage to check the best value of timeout for simple cases like bold button (or other not-dropdown buttons)? We have several issues here and we have to solve them one by one, so let's start with this simple case.

It seems that the default value introduced by this PR works fine in the most basic cases.

**Comment by mmotyczynska** [comment]
@ckeditor/qa-team I changed the fix completely. There is no timeout anymore so it can be tested on any Mac. Please kindly verify if all the issues mentioned earlier are solved now in Safari by this fix.

**Comment by godai78** [comment]
Recreated the problem in this branch:

**Comment by LukaszGudel** [comment]
I'm testing changes on this branch and it looks really good so far. I didn't found any places where the toolbar would unexpectedly disappear or focus would behave differently. Still, I will test these changes for some time to make sure that everything look good.

---

> In Safari, go to http://fake.ckeditor.com:8888/cke5-workspace/ckeditor5/build/docs/ckeditor5/34.2.0/examples/builds-custom/full-featured-editor.html
> Focus the content.
> Try using any toolbar item (with different results).

@godai78 I'm not able to recreate this issue on this branch. Are you on the latest commit `a59e996`?

**Comment by godai78** [comment]
> Are you on the latest commit `a59e996`?

I was convinced about it, but now I am not. Now I am testing this once again.

**Review by LukaszGudel — APPROVED** [review]

**Comment by Mgsy** [comment]
Unfortunately, I can still reproduce the disappearing toolbar issue with the media embed dropdown 😞 

https://user-images.githubusercontent.com/20988892/181271136-15789d0d-9c83-407f-afdf-33cc3a2d1e22.mp4

**Comment by mlewand** [comment]
> Unfortunately, I can still reproduce the disappearing toolbar issue with the media embed dropdown (only in Safari) :disappointed:

This should be fixed by Darek's fix - shouldn't it? (#12162)

File: packages/ckeditor5-ui/src/button/buttonview.js [hunk 1]
```diff
@@ -163,7 +164,17 @@ export default class ButtonView extends View {
 					}
 				} )
 			}
-		} );
+		};
+
+		// For safari we have to force the focus as it is implemented differently than in other browsers. See #12115.
```
Comment (Reinmar) [inline packages/ckeditor5-ui/src/button/buttonview.js:169]:
```suggestion
		// On Safari we have to force the focus on a button on click as it's the only browser
		// that doesn't do that automatically. See #12115.
```

**Comment by Reinmar** [comment]
Code-wise the change looks good to me too.

Just some inspiration for manual tests:

*   iOS
*   different editor types
*   clicking in different panels and buttons and inputs
*   clicking on the toolbar itself (between buttons)

At least the last bit wasn't working (deserves a followup, not a blocker).

**Comment by Mgsy** [comment]
> > Unfortunately, I can still reproduce the disappearing toolbar issue with the media embed dropdown (only in Safari) :disappointed:
> 
> This should be fixed by Darek's fix - shouldn't it? ([#12162](https://github.com/ckeditor/ckeditor5/pull/12162))

Unfortunately, it doesn't fix it. The media embed button is the last one that fails in Safari :/

**Comment by mmotyczynska** [comment]
> The media embed button is the last one that fails in Safari :/

A followup issue for this one #12171.

**Comment by mmotyczynska** [comment]
> At least the toolbar clicking wasn't working (deserves a followup, not a blocker, broken on master too).

A followup issue for this one #12172.

**Comment by mlewand** [comment]
Let's do a final QA testing, that includes:

*   CF features
*   iOS
*   Strong focus on split buttons.

The CKE5 itself has been already well covered by @Mgsy so he don't need to do this particular part that much (contrary to @FilipTokarski 😈).

Let's test `focus-release-hot-fixes` branch that contains both fixes (#12134, #12162).

**Review by Mgsy — APPROVED** [review]

**Review by FilipTokarski — APPROVED** [review]

**Pull request merged**